### ISSUE TITLE
Library doesn't record

### ISSUE THREAD

**Issue by dajver** [republicofgavin/PauseResumeAudioRecorder#2]
I start recording like this

String audio = Environment.getExternalStorageDirectory() + "/record_" + recordsListView.getSectionType();
        mediaRecorder = new PauseResumeAudioRecorder();
        mediaRecorder.setAudioFile(audio);
        mediaRecorder.startRecording();

and then I trying to listen it in VLC player, and it' doesn't work... What problem could be?

**Comment by dajver** [comment]
and i got start() status -38 error

**Comment by HRobbie** [comment]
May be the following questions:
1. Please check whether to add the tape permission.
2. Whether the tape resources occupied.
If not release resources after recording for the first time, will lead to the next recording was unable to complete.You can restart the phone and test again.

**Comment by dajver** [comment]
Thanks for your reply. But in some phones recording working and on some not. For example I tested this on Nexus 4 and Nexus 5 and recording doesnt work there. And on Meizu m3s and U10 works well... What it could be? Everywhere Android 5.1 instead of Nexus 5, there Android 6.

And on first start I got this error on Nexus phones.

I have all premissions what library need
android:name="android.permission.RECORD_AUDIO"
android:name="android.permission.WRITE_EXTERNAL_STORAGE"
android:name="android.permission.READ_EXTERNAL_STORAGE"

Thanks for your time.

**Comment by dajver** [comment]
Hey, are you here? :)

**Comment by republicofgavin** [comment]
Yeah, I'm here. Sorry for the delay. Busy week and all that. 
Under the covers, the system creates an AudioRecord(https://developer.android.com/reference/android/media/AudioRecord.html) to pull audio information from the peripheral. If the settings are wrong (like you try to record in stereo when you don't have stereo recording hardware or your PCM bit settings are wrong, the AudioRecord will not pick up any data.
For the phones you are trying to test, sorry I don't have those versions to test with you, can you run the sample app to see if it its configurations work for you? I tried to initialize the recorder with settings that work on all types of phones, but I may have failed in doing so.
If you can do that and have the sample app play the recording and you hear something, that tells me at least data is getting put into the file. Which means VLC may be having trouble playing it (long shot I know).

**Comment by dajver** [comment]
Thanks for your reply. Your example works well, but I need a bit another function. How could I record different audio files in one session? I tryed create each time new object of PauseResumeAudioRecorder on click but it doesnt work...

**Comment by dajver** [comment]
for example I want to do something like this

```
@OnClick(R.id.record)
    public void onRecordClick() {
           mediaRecorder = new PauseResumeAudioRecorder();
           mediaRecorder.setAudioFile(audioPath);
           mediaRecorder.startRecording();
}
```

audioPath - could be different with different names of files each click, but when I do like this, I got start() status -38 error in some devices...

How can I do this, could you help please?

**Comment by dajver** [comment]
How can I create new instance to create new file with new record? In same session in fragment?

Maybe you have some idea to change something in code? That fix this problem? Cuz I realy need it :D

I have debuged your code and found some problem. On meizu m3s this line of code works well, lenght = 2048
```

final int length = currentAudioRecording.read(readingBuffer, 0, bufferSizeInBytes);
                        for (int i = 0; i < length; i++) {
                            dataOutputStream.writeShort(readingBuffer[i]);
                        }
```

and on nexus 4 and 5 lenght = -3 and "for" doesn't do his work... Maybe it will help you to fix this issue?

**Comment by republicofgavin** [comment]
With my recording app, I force the recorder to be stopped before a new one is created. I don't think two recorders can operate (even if one or both are paused) at the same time, because they are both trying to pull data from the same hardware.
Thank you for debugging the code, I don't have access to those kinds of phones, so I can't really test.
I did notice this 
https://developer.android.com/reference/android/media/AudioRecord.html#read(java.nio.ByteBuffer,%20int)
Basically, -3 is the error code for "ERROR_INVALID_OPERATION if the object isn't properly initialized"

**Comment by republicofgavin** [comment]
I'm continuing to look into it, I just wanted to give you a heads up on my investigation so far

**Comment by republicofgavin** [comment]
So, I've been reading the AudioRecord source code trying to figure out why you are getting ERROR_INVALID_OPERATION on your read.
It is returned when the object has failed to initialize(tracked with a variable called mState) or you are using AudioFormat.ENCODING_PCM_FLOAT.
The mState variable gets set during only 2 operations in the code: during the constructor and during the release function.
This code caught my eye: https://github.com/android/platform_frameworks_base/blob/master/media/java/android/media/AudioRecord.java#L381
A todo comment is made (I can't dig past the native_setup call),  that refers to the capture device already being open. The following line just checks to see if it was successful, only if it is does that variable get set to initialized.

With that being said, Audio records cannot start without that variable being set (https://github.com/android/platform_frameworks_base/blob/master/media/java/android/media/AudioRecord.java#L974).

So, my only conjecture at this point is that the PauseResumeAudioRecorder was initialized, started, released, and then maybe started again. The current version of PauseResumeAudioRecorder doesn't let the consumer start recording if stop has been called. It will display a log saying the operation is not allowed(Audio recorder is not in prepared state. Ignoring call.). Based off of your code snippet, you seem to be using the recorder correctly, so I'm at loss, if my conjecture is wrong. :/
So, a few questions:
1. In your code, did you extend the recorder to provide new functionality or is it the version in github?
2. Are you calling stop on the previous recorder (if it exists and is recording) before you create a new instance of one and start recording?
3. Can you get me a log snippet?

Sorry for your inconvenience.

**Comment by republicofgavin** [comment]
Something else you could try to solve your problem is cycling through PCM and sample rates. Not every device supports every combination of settings (http://stackoverflow.com/questions/19517600/android-audio-recorder-initialized-failed). If you find a combination that works on the nexus 4 and nexus 5, then you are good to go. I noticed that I didn't release the audio record in my thread (even though that hasn't caused problems for my recorder. I'm going to update the code to do that and increment the version number (assuming it works).

**Comment by dajver** [comment]
Thank you very much for your response, It is so full-dress. So I will answer on your questions:

1. I'm using github version of code, I didn't change it.
2. Yes, after each time when user call play button, this button became disable and user can't click on it till the moment when he will click on stop.
3. What log snippet do you need? Here my example how I try to record:

Start method, as you see I trying to create object of PauseResumeAudioRecorder on click method. 
audioId - is just int with some integer value.
bundle - is just needed for get state of what todo, create new or use old folder. Don't think that it could make problems...

```
@OnClick(R.id.record)
    public void onRecordClick() {
        if(bundle == null) {
            File postcastDerictory = new File(Environment.getExternalStorageDirectory() + "/PosterCast/" + audioId + "/");
            postcastDerictory.mkdirs();
            audioPath = postcastDerictory.getAbsolutePath() + "/record_" + recordsListView.getSectionType();
        } else {
            audioPath = Environment.getExternalStorageDirectory() + "/PosterCast/" + audioId + "/record_" + recordsListView.getSectionType();
        }
        mediaRecorder = new PauseResumeAudioRecorder();
        mediaRecorder.setAudioFile(audioPath);
        mediaRecorder.startRecording();

        record.setVisibility(View.GONE);
        pause.setVisibility(View.VISIBLE);
        stop.setEnabled(true);
        next.setEnabled(false);
    }
```

Here is my stop method:

```
@OnClick(R.id.stop)
    public void onStopClick() {
        mediaRecorder.stopRecording();

        record.setVisibility(View.VISIBLE);
        pause.setVisibility(View.GONE);
        resume.setVisibility(View.GONE);
        stop.setEnabled(false);
        next.setEnabled(true);
    }
```

Here id my Pause method:

```
@OnClick(R.id.pause)
    public void onPauseClick() {
        mediaRecorder.pauseRecording();

        resume.setVisibility(View.VISIBLE);
        pause.setVisibility(View.GONE);
        stop.setEnabled(false);
        next.setEnabled(false);
    }
```

Here is my resume method:

```
@OnClick(R.id.stop)
    public void onStopClick() {
        mediaRecorder.stopRecording();

        record.setVisibility(View.VISIBLE);
        pause.setVisibility(View.GONE);
        resume.setVisibility(View.GONE);
        stop.setEnabled(false);
        next.setEnabled(true);
    }
```

And onDestroy method:

```
@Override
    public void onDestroy() {
        super.onDestroy();
        if (mediaRecorder.getCurrentState() == PauseResumeAudioRecorder.RECORDING_STATE || mediaRecorder.getCurrentState() == PauseResumeAudioRecorder.PAUSED_STATE)
                mediaRecorder.stopRecording();
    }
```

Maybe problem because of that I trying create object of PauseResumeAudioRecorder on click ? If that, how I can create multiply of audio files in one activity?

**Comment by dajver** [comment]
And you don't have release method, I undestand that you have it in stop method, but maybe it doesn't work... Or I don't know. App doesn't record even on first start, on first click when I try to initialize on click...

**Comment by dajver** [comment]
> So, my only conjecture at this point is that the PauseResumeAudioRecorder was initialized, started, released, and then maybe started again

I doing like that. And I want to do like this more than 10 times. I want to record 12 audio records in a row... So I create new object of library and doing start, after recording of some seconds of audio I click stop, than again creating new object of library and again start, stop with new path and file name and so on. So can't I do like this?

And why it works on meizu and xiaomi and doesn't work on nexus and LG devices. I confused :D

**Comment by dajver** [comment]
I found solution for me :) In my devices (nexus's) thread started twice and that is why problem appear. I started to check if thread exist to not create second and it helped me. Thanks for your help man, you are great developer!

**Comment by republicofgavin** [comment]
Thank you for the compliment and thank you for using the recorder. I am glad your problem is solved! Feel free to contact me again if you need.
